ports/psoc6/machine_pwm.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2022-2024 Infineon Technologies AG
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "modmachine.h"
+
+// port-specific includes
+#include "machine_pin_phy.h"
+#include "mplogger.h"
+
+typedef struct _machine_pwm_obj_t {
+    mp_obj_base_t base;
+    cyhal_pwm_t pwm_obj;
+    uint32_t pin;
+    uint32_t fz;
+    uint8_t duty_type;
+    mp_int_t duty;
+    // bool invert;
+} machine_pwm_obj_t;
+
+static machine_pwm_obj_t *pwm_obj[MAX_PWM_OBJS] = { NULL };
+
+static inline machine_pwm_obj_t *pwm_obj_alloc() {
+    for (uint8_t i = 0; i < MAX_PWM_OBJS; i++)
+    {
+        if (pwm_obj[i] == NULL) {
+            pwm_obj[i] = mp_obj_malloc(machine_pwm_obj_t, &machine_pwm_type);
+            return pwm_obj[i];
+        }
+    }
+
+    return NULL;
+}
+
+static inline void pwm_obj_free(machine_pwm_obj_t *pwm_obj_ptr) {
+    for (uint8_t i = 0; i < MAX_PWM_OBJS; i++)
+    {
+        if (pwm_obj[i] == pwm_obj_ptr) {
+            pwm_obj[i] = NULL;
+        }
+    }
+}
+
+enum {
+    VALUE_NOT_SET = -1,
+    DUTY_NOT_SET = 0,
+    DUTY_U16,
+    DUTY_NS
+};
+
+static void mp_machine_pwm_freq_set(machine_pwm_obj_t *self, mp_int_t freq);
+
+static cy_rslt_t pwm_freq_duty_set(cyhal_pwm_t *pwm_obj, uint32_t fz, float duty_cycle) {
+    return cyhal_pwm_set_duty_cycle(pwm_obj, duty_cycle * 100, fz); // duty_cycle in percentage
+}
+
+static inline cy_rslt_t pwm_duty_set_ns(cyhal_pwm_t *pwm_obj, uint32_t fz, uint32_t pulse_width) {
+    return cyhal_pwm_set_period(pwm_obj, 1000000 / fz, pulse_width / 1000); // !# * --> /
+}
```

Comment:
It may be more convenient to choose one of these functions `cyhal_pwm_set_duty_cycle` or `cyhal_pwm_set_period` and use only that one in both cases.

Also, it is possible to recalculate a duty at the last minute before calling `cyhal_pwm_set_XXX` and remove `float` and `double` in other places.

Reply:
Understand it might be a bit confusing. But, taking reference from here: https://ifx-micropython.readthedocs.io/en/latest/library/machine.PWM.html#machine.PWM

1. When duty_u16 is set, it is set as raw value and `cyhal_pwm_set_duty_cycle` allows to pass duty_cycle as percentage. 
2. When duty_ns is set, it is set as time in ns and `cyhal_pwm_set_period` allows to directly pass time. 

It helps to reuse the low-level API's as much as intended functionality is provided through them. Does that explains?

Reply:
Yes. 
Anyhow, I think as long as possible is a good idea to restrict the conversion to one place. @IhorNehrutsa let us check if we can refactor without breaking any functionality :)

Reply:
As mentioned in the other comment:
https://github.com/Infineon/micropython/commit/3c14c39b7238df657a0be46c6ec46cd1d95d036f

I have kept the 2 ways of setting freq/duty. But now abstracted and encapsulated under `pwm_config()`.